programs/marinade-finance/src/stake_system/partial_unstake.rs

```diff
@@ -0,0 +1,293 @@
+        self.state.stake_system.check_stake_list(&self.stake_list)?;
+        self.state
+            .check_stake_deposit_authority(self.stake_deposit_authority.key)?;
+        check_owner_program(&self.stake_account, &stake::program::ID, "stake_account")?;
+        self.state
+            .check_stake_deposit_authority(self.stake_deposit_authority.key)?;
+        check_address(self.stake_program.key, &stake::program::ID, "stake_program")?;
+
+        let mut stake = self
+            .state
+            .stake_system
+            .get(&self.stake_list.data.as_ref().borrow(), stake_index)?;
+        if self.stake_account.to_account_info().key != &stake.stake_account {
+            msg!(
+                "Stake account {} must match stake_list[{}] = {}. Maybe list layout was changed",
+                self.stake_account.to_account_info().key,
+                stake_index,
+                &stake.stake_account
+            );
+            return Err(ProgramError::InvalidAccountData);
+        }
+        assert!(
+            stake.is_emergency_unstaking == 0,
+            "already emergency unstaking"
+        );
+
+        let mut validator = self
+            .state
+            .validator_system
+            .get(&self.validator_list.data.as_ref().borrow(), validator_index)?;
+
+        // check that the account is delegated to the right validator
+        check_stake_matches_validator(&self.stake_account.inner, &validator.validator_account)?;
+
+        // Allow partial unstake
+        // compute target for this particular validator (total_active_balance * score/total_score)
+        // Note: we use total_active_balance without considering stake_delta, because stake_delta can change before the stake_delta window
+        let total_stake_target = self.state.validator_system.total_active_balance;
+
+        let validator_stake_target = self
+            .state
+            .validator_system
+            .validator_stake_target(&validator, total_stake_target)?;
+
+        // compute how much we should unstake from this validator
+        if validator.active_balance <= validator_stake_target {
+            msg!(
+                "Validator {} has already reached unstake target {}",
+                validator.validator_account,
+                validator_stake_target
+            );
+            return Ok(()); // Not an error. Don't fail other instructions in tx
+        }
+        let unstake_from_validator = validator.active_balance - validator_stake_target;
+        msg!(
+            "partial unstake {} from_validator {}",
+            unstake_from_validator,
+            &validator.validator_account
+        );
+
+        // compute how much this particular account should have
+        let stake_account_target = stake
+            .last_update_delegated_lamports
+            .saturating_sub(unstake_from_validator);
+
+        let unstake_amount = if stake_account_target < 2 * self.state.stake_system.min_stake {
```

Comment:
Unstake amount must be a parameter for this function. No whole stake deactivation here (use emergency unstake instead). We need just check the input unstake amount to not go below the target level and to not make one of resulting stakes below the minimum stake amount. In case of violation just return an error